### DISCUSSION TITLE
Custom RunTask activity

### DISCUSSION THREAD

**Discussion by anghelnicolae** [elsa-workflows/elsa-core#6090] [Q&A]
Good morning,
I am trying to create a custom activity that works similarly to the RunTask activity. I've basically copied the code from RunTask and the webhook is called, but the completion endpoint only works for RunTask activities. Am I missing something or is this a limitation of the platform?
Can I use the MassTransit integration to do something similar?

**Comment by sfmskywalker** [comment]
Indeed, the completion endpoint is designed only to target the RunTask activity. Currently, none of the trigger activities can be invoked without specifying their type directly. Until we have that in place, you'd have to also copy the code for the `ITaskReporter` and its implementation in order to trigger your custom activity.

**Reply by anghelnicolae** [reply]
Hi! Thank you for the answer. Is there an easy way to offload activity execution to multiple remote systems? Adding Activity execution to the MassTransit integration would be really helpful. From what I understand it can only be used to start workflows now.

**Reply by anghelnicolae** [reply]
Just noticed that there was something like this in Elsa 2, but hasn't been migrated to Elsa 3.

**Reply by sfmskywalker** [reply]
Can you elaborate on what you mean specifically with offloading activity execution to multiple remote systems?

**Reply by anghelnicolae** [reply]
> Can you elaborate on what you mean specifically with offloading activity execution to multiple remote systems?

My objective is to have a workflow server that sends tasks to workers on other computers. Once the workers finish a task, the workflow server should be notified to resume the workflow. Could you recommend the simplest approach to achieve this?

**Comment by rolandgriesser** [comment]
Hi @anghelnicolae , I had a similar use-case. I'm using the following code, it works fine for me (because of the `activityTypeName`). 

```
public class ExtendedRunTask : Activity<object>
{
    /// <summary>
    /// The key that is used for sending and receiving activity input.
    /// </summary>
    public const string InputKey = "RunTaskInput";

    [Port]
    public IActivity? TaskIdGenerated { get; set; }

    protected override async ValueTask ExecuteAsync(ActivityExecutionContext context)
    {
        var identityGenerator = context.GetRequiredService<IIdentityGenerator>();
        var taskId = identityGenerator.GenerateId();

        var payload = new RunTaskStimulus(taskId, nameof(ExtendedRunTask));
        var activityTypeName = ActivityTypeNameHelper.GenerateTypeName<RunTask>();
        context.CreateBookmark(new CreateBookmarkArgs()
        {
            BookmarkName = activityTypeName,
            Callback = ResumeAsync,
            IncludeActivityInstanceId = false,
            Stimulus = payload
        });

        if (TaskIdGenerated != null)
        {
            var variables = new[]
            {
                new Variable("TaskId", taskId),
            };
            await context.ScheduleActivityAsync(TaskIdGenerated, null, variables: variables);
        }
    }

    private async ValueTask ResumeAsync(ActivityExecutionContext context)
    {
        var input = context.GetWorkflowInput<object>(InputKey);

        context.Set(Result, input.ToExpandoObject());

        var outcomes = new List<string>
        {
            "Done"
        };

        await context.CompleteActivityWithOutcomesAsync(outcomes.ToArray());
    }
}
```

My use case is the following - I need the task id for an external system, so I call the port `TaskIdGenerated`. Within that port, I can then send an email with the callback-url or create a task in an external system. And with that link, I can continue the workflow with the `Done` outcome.

![image](https://github.com/user-attachments/assets/bd3c7dd7-1ff0-4b9d-a180-0db7370f264e)

**Reply by anghelnicolae** [reply]
> activityTypeName

Hi @rolandgriesser, sorry for the very late reply, but I'm just starting looking into this again. Could you tell me more about ActivityTypeNameHelper.GenerateTypeName, does it have to be a unique value for each instance. Could you also remind me what the resume URL format looks like? Thank you.

**Comment by StevenMcCawley** [comment]
@anghelnicolae if I'm understanding your issue correctly, check out my work-around

https://github.com/elsa-workflows/elsa-core/issues/7068

**Comment by sfmskywalker** [comment]
Short answer: yes, the current `/tasks/{taskId}/complete` endpoint is specific to `RunTask`.

Relevant source:
- `elsa-core/src/modules/Elsa.Workflows.Api/Endpoints/Tasks/Complete/Endpoint.cs:7-10,28-31` explicitly says it resumes the `RunTask` activity and forwards completion to `ITaskReporter`.
- `elsa-core/src/modules/Elsa.Workflows.Runtime/Services/TaskReporter.cs:11-26` hard-codes `ActivityTypeNameHelper.GenerateTypeName<RunTask>()`.
- `elsa-core/src/modules/Elsa.Workflows.Runtime/Activities/RunTask.cs:96-115` creates the bookmark from `RunTaskStimulus(taskId, taskName)` and resumes by reading `RunTask.InputKey`.

So for your follow-up questions:
- `ActivityTypeNameHelper.GenerateTypeName<T>()` is not unique per instance. It is the type name for the activity type.
- Per-instance uniqueness comes from the `taskId`. `RunTaskStimulus` excludes `TaskName` from the hash, so the task ID is the important part for matching (`elsa-core/src/modules/Elsa.Workflows.Runtime/Stimuli/RunTaskStimulus.cs:9`).
- The built-in `/tasks/{taskId}/complete` endpoint will not resume a custom `RunTask`-like activity unless you also provide a custom reporter/endpoint (or a genericized version of `ITaskReporter`).

There is already a related enhancement request here: #7068.

If your goal is remote workers on other machines, the most supported pattern in the current codebase is still:
1. use the built-in `RunTask`,
2. handle the emitted `RunTaskRequest`, and
3. complete it through `ITaskReporter`.

The sample `elsa-samples/src/aspnet/Elsa.Samples.AspNet.RunTaskIntegration` is the best reference for that shape.

Closing this discussion as answered, with #7068 tracking the generic custom-activity enhancement.
